ens-improvement-proposals/ensip-15-ccip-read-metadata.md

```diff
@@ -0,0 +1,137 @@
+description: Allows metadata to be queried on CCIP enabled names
+---
+
+# ENSIP-XX: CCIP-read metadata
+
+| **Author**    | Jeff Lau \<jeff@ens.domains> |
+| ------------- | ---------------------------- |
+| **Status**    | Draft                        |
+| **Submitted** | 2022-09-22                   |
+
+### Abstract
+
+This ENSIP allows metadata to be queried directly on the resolver for CCIP-read enabled names. CCIP-read will power many of the domains in the future, however since the retrieval mechanism uses wildcard + offchain resolver, there is no standardised way to retrieve important metadata information such as the owner (who can change the records), or which L2/offchain database the records are stored on.
+
+### Motivation
+
+With CCIP-read subdomains already starting to get used by larger partners in the ENS ecosystem, it is important that there is a way of frontend interfaces to get important metadata to allow a smooth user interface. For instance the owner of a CCIP enabled name would need to be known, for a UI to show whether or not the currently connected account has the rights to edit the records. If the owner was not known, the only way to know would be to make the edit and wait for failure.
+
+This ENSIP addresses this by adding a way of important metadata to be gathered on the offchain resolver, which would likely revert and be also resolved offchain, however there is an option for it to be also left onchain if there value was static and wouldn't need to be changed often.
+
+### Specification
+
+The metadata should include 2 different types of info
+
+- Offchain data storage location related info: `graphqlUrl` includes the URL to fetch the metadata.
+
+- Ownership related info: `owner`, `isApprovedForAll` defines who can own or update the given record. The specification also must comply with [OwnedNode](https://github.com/corpus-io/Optimism-Resolver/blob/main/contracts/l2/L2PublicResolver.sol) that consists of the hash of the node and `msg.sender`. The OwnedNode is used when the resolver does not depend on the canonical registry nor a bridge to verify the ownership of the name on L1.
+
+```solidity
+
+interface OffChainResolver {
+    /** @dev Returns the owner of the resolver on L2
+     * @param node
+     * @return owner gives space for longer address on non-evm chains.
+     */
+    owner(bytes32 node) returns (bytes32 owner);
+
+    isApprovedForAll(address account, address operator) returns (boolean);
+    /** @dev Returns the owner of the resolver on L2
+     * @return name can be l2 chain name or url if offchain
+     * @return coinType according to https://github.com/ensdomains/address-encoder
+     * @return graphqlUrl url of graphql endpoint that provides additional information about the offchain name and its subdomains
+     * @return storageType 0 = EVM, 1 = Non blockchain, 2 = Starknet
+     * @return context can be l2 resolver contract address for evm chain but can be any l2 storage identifier for non evm chain
```

Comment:
I had a discussion with @Arachnid and would like to change the following

- @return context = This returns the same data used as `context` when calling `addr(bytes calldata context, bytes32 node, uint256 coinType)` at https://github.com/corpus-io/ENS-Bedrock-Resolver/blob/main/contracts/l2/profiles/addr/AddrResolver.sol#L98 (in your specific implementation, it needs to return the owner address of the name).
- @return storageLocation = This is where it returns the l2 resolver contract address for evm chain but can be any l2 storage identifier for non evm chain

Reply:
@AlexNi245 can you add storage location? Also can you recreate this PR against `ensip-14` branch rather than against `master`?